src/mmfatl.h

```diff
@@ -10,74 +10,305 @@
+ *   unnecessary correction under severe conditions.
+ *   
+ * 
+ * The \p format is followed by a possibly empty list of paramaters substituted
+ *   for placeholders.  Currently unsigned int values may replace a
+ *   \ref MMFATL_PH_UNSIGNED type placeholder, and a char const* pointer a
+ *   \ref MMFATL_PH_STRING type placeholder.  If the latter pointer is NULL,
+ *   the placeholder is replaced with an empty string, else it must point to
+ *   ASCII encoded NUL terminated text.  No value is required for the
+ *   \ref MMFATL_PH_PREFIX type tokens.
+ * \return false iff the message buffer is in overflow state.
+ * 
+ * \pre \p format if not NULL, contains NUL terminated ASCII text.
+ * \pre \ref fatalErrorInit was called before.
+ * \pre the submitted parameters following \p format must match in type and
+ *   order the placeholders in \p format.  Their count may exceed that of the
+ *   placeholders, but must never be less.
+ * \post the message is appended to the current buffer contents.  It is
+ *   truncated if there is insufficient space for it, including the
+ *   terminating NUL.
+ * \invariant the memory state of the rest of the program is not changed.
+ * \warning This function does not automatically wrap messages and insert LF
+ *   characters at the end of the declared screen width (g_screenWidth in
+ *   mminou.h) as print2 does.  Tests show that usual terminal emulators break
+ *   up text at the last column, but that may depend on the used
+ *   hard-/software.
+ */
+extern bool fatalErrorPush(char const* format, ...);
+
+/*!
+ * \brief display buffer contents and exit program with code EXIT_FAILURE.
+ * 
+ * This function does not return.
+ *
+ * A NUL terminated message has been prepared in an internal buffer using a
+ * sequence of \ref fatalErrorPush.  This function writes this message to
+ * stderr (by default tied to the terminal screen like stdout), and exits the
+ * program afterwards, indicating a failure to the operating system.
+ *
+ * A line feed is appended to a prepared non-empty message if it is not its
+ * last character.  This keeps the message and a command prompt following the
+ * exit on separare lines.  It is recommended to avoid this corrective
+ * measure and supply a terminating line feed as part of the message.
+ *
+ * It is possible to call this function without preparing a message.  In this
+ * case nothing is written to stderr, and the program just exits with a failure
+ * code.
+ *
+ * \pre \ref fatalErrorInit has initialized the internal error message
+ *   buffer, possibly followed by a sequence of \ref fatalErrorPush
+ *   filling it with a message.
+ * \post [noreturn] the program terminates with error code EXIT_FAILURE, after
+ *   writing the buffer contents to stderr.
+ * \post a line feed is appended to any non-empty message, if it is not
+ *   provided
+ * \invariant the memory state of the rest of the program is not changed.
```

Comment:
```suggestion
```
This line doesn't make any sense, the function doesn't return so there is no state to return to.

Reply:
It does make a lot of sense, since a function may have been registered with atexit https://en.cppreference.com/w/c/program/atexit.  Such a function is executed after exit was called, i.e. after the code here is finished.  Usually you deallocate resources there, and so such function needs access to data describing them.  Why should a fatal error handler destroy such data? Because an implementer might want to hard reset the heap just to free some space in a memory tight situation, for running printf instead of my poor man's implementation here, that is required by the specs of the C library.
Your comment here underlines the necessity of keeping the invariant.  The above explanation is found in two locations.  I did not copy it to all invariant lines, though.

Reply:
I see. That is not a situation that occurs in the program, so I think it is safe not to consider it, but if you would like to document it anyway then you should make it clear that the invariant is talking about atexit handlers. I think the conventional reading of invariants is that they relate the program state before and after a call.